### PR TITLE
chore(tests): Allow compiling examples and tests with stable toolchain

### PR DESCRIPTION
**Pull request by ac-freeman**
As in #48, this PR allows the library to work on the `stable` and `beta` Rust channels by removing the `associated_type_defaults` and `never_type` features.

### CODE REVIEW

**Comment by danieleades** [comment]
Hi @ac-freeman, thanks for the contribution.

I think this PR goes a little further than the stated goal of removing the dependency on the nightly compiler, so maybe you can talk me through some of the changes.

Much of this hangs on the following question-  Is the goal to remove the dependency on the nightly compiler so that downstream users of the crate do not need to use nightly, or so that contributors do not need to use the nightly compiler to run tests and examples?

- If the former then removing the associated type defaults should be sufficient
   - the 'never' type is only used in doc comments and examples, and not in the library itself. It shouldn't be necessary to remove this in order to allow downstream users to compile with a stable toolchain.
   - a CI job would need to be added to confirm that the crate (sans examples and tests) could be compiled with the stable toolchain
 - you've modified some `Error`s from generics to boxed traits. What is the reason for this?

**Comment by danieleades** [comment]
see https://github.com/danieleades/arithmetic-coding/pull/64

**Comment by ac-freeman** [comment]
> Is the goal to remove the dependency on the nightly compiler so that downstream users of the crate do not need to use nightly, or so that contributors do not need to use the nightly compiler to run tests and examples?

Yes, both, although primarily the former. 

I think it's useful, however, if the examples can be run without the use of nightly. A user may want to get hands-on time by trying out the example code before integrating the library in their own project. Having to use nightly seems like a roadblock. Is there a reason at this point to prefer `!` over `Infallible`, when we don't know when (or if) `!` will be stabilized?

> you've modified some Errors from generics to boxed traits. What is the reason for this?

That was chasing down a red herring from a `thiserror` macro I had added--my apologies. It's been removed.


If you at least agree that the changes in #64 will be helpful, I think that's a great start, and it will make my life easier (updating Rust nightly has a habit of breaking other packages that I depend on).

**Comment by danieleades** [comment]
thanks @ac-freeman
I'm going to merge the other PR which removes the need for a nightly compiler for downstream users of the crate. That will then reduce the scope of this PR to removing the need for contributors or for people making direct use of the examples to use a nightly compiler. Apologies in advance for the rebasing!

File: Cargo.toml [hunk 1]
```diff
@@ -44,4 +44,4 @@ name = "sherlock"
 harness = false
 
 [lints]
-workspace = true
+workspace = true
```
Comment (danieleades) [inline Cargo.toml:47]:
```suggestion
workspace = true

```

File: fenwick-model/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,8 @@ categories.workspace = true
 repository.workspace = true
 rust-version.workspace = true
 
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
```
Comment (danieleades) [inline fenwick-model/Cargo.toml:19]:
```suggestion
```

**Review by danieleades — CHANGES_REQUESTED** [review]
minor nits

**Pull request merged**